TargetEnvironmentDataModel.md

```diff
@@ -0,0 +1,194 @@
+
+
+```sql
+-- "Base" asterism table.  Here we make the primary key be a combination of
+-- (asterism_id, asterism_type, instrument).  This is done so that the sub-
+-- tables can have a foreign key reference to asterism that includes all three,
+-- ensuring that they match.
+
+CREATE TABLE asterism (
+  asterism_id    SERIAL,
+  asterism_type  asterism_type NOT NULL,
+  instrument     identifier    NOT NULL    REFERENCES e_instrument,
+  PRIMARY KEY (asterism_id, asterism_type, instrument)
+);
+```
+
+Now we update observation to reference its one and only asterism.
+
+```sql
+-- Add columns to observation to track the associated asterism.  Along with
+-- instrument, these become a foreign key pointing into the base asterism
+-- table.
+
+ALTER TABLE observation
+  ADD asterism_id integer         NOT NULL,
+  ADD asterism_type asterism_type NOT NULL;
+
+-- Add a foreign key reference to the corresponding asterism. This ensures
+-- that the type and instrument correspond to the observation.
+ALTER TABLE observation
+  ADD FOREIGN KEY (asterism_id, asterism_type, instrument) REFERENCES asterism ON DELETE CASCADE;
+
+-- Ensure that no asterism appears in multiple observations.
+ALTER TABLE observation
+  ADD CONSTRAINT unique_asterism_per_obs UNIQUE (asterism_id);
+```
+
+Finally, we create the specific asterism sub-tables.  For the `single_target_asterism` any instrument except GHOST may be used.
+
+```sql
+-- Single target asterism can have any instrument except GHOST but it should
+-- match the "base" and observation.
+
+CREATE TABLE single_target_asterism (
+  asterism_id   integer       NOT NULL,
+  asterism_type asterism_type NOT NULL,
+  instrument    identifier    NOT NULL,
+  target_id     integer       NOT NULL  REFERENCES target(target_id),
+  PRIMARY KEY (asterism_id, asterism_type, instrument) REFERENCES asterism ON DELETE CASCADE,
+  CONSTRAINT is_single CHECK (asterism_type = 'SingleTarget'),
+  CONSTRAINT is_not_ghost CHECK (instrument != 'GHOST')
+);
+
+
+-- A placeholder dual target asterism, just to have a second instance of
+-- asterism.
+
+CREATE TABLE ghost_dual_target_asterism (
+  asterism_id   integer       NOT NULL,
+  asterism_type asterism_type NOT NULL,
+  instrument    identifier    NOT NULL,
+  target_id_1   integer       NOT NULL  REFERENCES target(target_id),
+  target_id_2   integer       NOT NULL  REFERENCES target(target_id),
+  -- more configuration here for guide fibers, etc.
+  PRIMARY KEY (asterism_id, asterism_type, instrument) REFERENCES asterism ON DELETE CASCADE,
+  CONSTRAINT is_ghost_dual (asterism_type = 'GhostDualTarget'),
+  CONSTRAINT is_ghost      (instrument    = 'GHOST')
+);
+```
+
+An observation will be deleted if its asterism is deleted, but we'll need a trigger to delete the asterism when the corresponding observation is deleted.
```

Comment:
I think we should not cascade here because the observation is logically the owner; deleting the asterism out from under an observation seems like a logic error. The trigger on the other end makes sense.